diff --git a/src/Plugin/GraphQL/DataProducer/DataProducerProxy.php b/src/Plugin/GraphQL/DataProducer/DataProducerProxy.php
index 489a3de..c555c1b 100644
--- a/src/Plugin/GraphQL/DataProducer/DataProducerProxy.php
+++ b/src/Plugin/GraphQL/DataProducer/DataProducerProxy.php
@@ -122,6 +122,14 @@ class DataProducerProxy implements ResolverInterface {
     return $this;
   }
 
+  /**
+   * Return map to this resolver.
+   * @return \Drupal\graphql\GraphQL\Resolver\ResolverInterface[]
+   */
+  public function getMap() {
+    return $this->mapping;
+  }
+
   /**
    * @param bool $cached
    *
