diff --git a/.htaccess b/.htaccess
index 246edc2..e579b6b 100644
--- a/.htaccess
+++ b/.htaccess
@@ -128,4 +128,8 @@ DirectoryIndex index.php index.html index.htm
       Header append Vary Accept-Encoding
     </FilesMatch>
   </IfModule>
+
+  # PHP FastCGI doesn't support HTTP Basic Authentication out of the
+  # box so we need this workaround. See http://drupal.org/node/1864628.
+  RewriteRule .* - [E=HTTP_AUTHORIZATION:%{HTTP:Authorization}]
 </IfModule>
